clouddriver-api/src/main/java/com/netflix/spinnaker/clouddriver/orchestration/AnnotationsBasedAtomicOperationsRegistry.java

```diff
@@ -0,0 +1,168 @@
+      if (cloudProvider == null) {
+        throw e;
+      }
+    }
+
+    // Operations can be versioned
+    VersionedDescription versionedDescription = VersionedDescription.from(description);
+
+    Class<? extends Annotation> providerAnnotationType = getCloudProviderAnnotation(cloudProvider);
+
+    List converters =
+        applicationContext.getBeansWithAnnotation(providerAnnotationType).values().stream()
+            .filter(
+                (it) -> {
+                  Object converterType =
+                      DefaultGroovyMethods.invokeMethod(
+                          it.getClass().getAnnotation(providerAnnotationType),
+                          "value",
+                          new Object[0]);
+                  VersionedDescription converterVersion =
+                      VersionedDescription.from((String) converterType);
+                  return converterVersion.descriptionName.equals(
+                          versionedDescription.descriptionName)
+                      && it instanceof AtomicOperationConverter;
+                })
+            .collect(Collectors.toList());
+
+    converters =
+        VersionedOperationHelper.findVersionMatches(versionedDescription.version, converters);
+
+    if (converters.isEmpty()) {
+      throw new AtomicOperationConverterNotFoundException(
+          String.format(
+              "No atomic operation converter found for description '%s' and cloud provider '%s'. It is possible that either 1) the account name used for the operation is incorrect, or 2) the account name used for the operation is unhealthy/unable to communicate with %s.",
+              description, cloudProvider, cloudProvider));
+    }
+    if (converters.size() > 1) {
+      throw new RuntimeException(
+          String.format(
+              "More than one (%s) atomic operation converters found for description '%s' and cloud provider '%s'",
+              converters.size(), description, cloudProvider));
+    }
+    return (AtomicOperationConverter) converters.get(0);
+  }
+
+  @Override
+  public DescriptionValidator getAtomicOperationDescriptionValidator(
+      String validator, String cloudProvider) {
+    // Legacy naming convention which is not generic and validator name is specific to cloud
+    // provider
+    try {
+      DescriptionValidator descriptionValidator =
+          super.getAtomicOperationDescriptionValidator(validator, cloudProvider);
+      if (descriptionValidator != null) {
+        return descriptionValidator;
+      }
+    } catch (NoSuchBeanDefinitionException e) {
+    }
+
+    if (cloudProvider == null) return null;
+
+    Class<? extends Annotation> providerAnnotationType = getCloudProviderAnnotation(cloudProvider);
+
+    List validators =
+        applicationContext.getBeansWithAnnotation(providerAnnotationType).values().stream()
+            .filter(
+                (it) -> {
+                  Object validatorType =
+                      DefaultGroovyMethods.invokeMethod(
```

Comment:
I don't think you need Groovy for this - should just be able to use Java reflection.